In [1]:
import os
import cv2
import random
import numpy as np
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report,accuracy_score

In [2]:
dir_train ='/Users/eslemus/Desktop/ML/Anticl/images/Train'
dir_test ='/Users/eslemus/Desktop/ML/Anticl/images/Test'
label_type = ['Class1','Class2']

img_size = 150
im_l =240
im_b =320
training_data = []

In [3]:
def create_data(dir_train,label_type):
    for seismic_type in label_type:
        path = os.path.join(dir_train,seismic_type)
        classnum = label_type.index(seismic_type)
        for img in os.listdir(path):
            img_arr = cv2.imread(os.path.join(path,img),0)
            print(img)
            if img_arr is not None:
            	new_arr = cv2.resize(img_arr,(img_size,img_size))
            	training_data.append([new_arr,classnum])
            
create_data(dir_train,label_type)
print(len(training_data))
random.shuffle(training_data)

63.jpg
189.jpg
77.jpg
162.jpg
176.jpg
88.jpg
200.jpg
201.jpg
177.jpg
89.jpg
163.jpg
188.jpg
76.jpg
62.jpg
74.jpg
60.jpg
48.jpg
149.jpg
175.jpg
161.jpg
203.jpg
202.jpg
160.jpg
174.jpg
148.jpg
49.jpg
61.jpg
75.jpg
59.jpg
71.jpg
65.jpg
164.jpg
158.jpg
206.jpg
207.jpg
159.jpg
165.jpg
171.jpg
64.jpg
70.jpg
58.jpg
198.jpg
66.jpg
72.jpg
167.jpg
99.jpg
173.jpg
205.jpg
204.jpg
210.jpg
172.jpg
166.jpg
98.jpg
73.jpg
67.jpg
9.jpg
14.jpg
._33.jpg
28.jpg
._27.jpg
129.jpg
101.jpg
115.jpg
114.jpg
100.jpg
128.jpg
29.jpg
15.jpg
17.jpg
116.jpg
102.jpg
103.jpg
117.jpg
12.jpg
113.jpg
107.jpg
106.jpg
112.jpg
13.jpg
39.jpg
11.jpg
104.jpg
138.jpg
111.jpg
105.jpg
10.jpg
38.jpg
21.jpg
35.jpg
108.jpg
120.jpg
134.jpg
._194.jpg
._6.jpg
135.jpg
121.jpg
109.jpg
20.jpg
36.jpg
22.jpg
137.jpg
123.jpg
122.jpg
136.jpg
37.jpg
33.jpg
27.jpg
132.jpg
126.jpg
127.jpg
133.jpg
26.jpg
32.jpg
18.jpg
24.jpg
30.jpg
125.jpg
131.jpg
119.jpg
._2.jpg
._146.jpg
118.jpg
130.jpg
124.jpg
31.jpg
25.jpg
19.jpg
42.jpg
4.jpg
56.jpg
180.jpg
194

In [4]:
x=[]
y=[]
for feature , label in training_data:
    x.append(feature)
    y.append(label)
print(len(x),len(y))
X = np.array(x)

print(X.shape)

400 400
(400, 150, 150)


In [5]:
train_imgs_scaled = X.astype('float32')
train_imgs_scaled /= 255

x_train_new = train_imgs_scaled.reshape(len(train_imgs_scaled),train_imgs_scaled.shape[1]*train_imgs_scaled.shape[2])

In [6]:
logisticRegr = LogisticRegression()
logisticRegr.fit(x_train_new, y)

create_data(dir_test,label_type)
print(len(training_data))
random.shuffle(training_data)

test_x=[]
test_y=[]

for feature , label in training_data:
    test_x.append(feature)
    test_y.append(label)
print(len(test_x),len(test_y))
tx = np.array(test_x)

print(X.shape)

test_imgs_scaled = tx.astype('float32')
test_imgs_scaled /= 255

x_test_new = test_imgs_scaled.reshape(len(test_imgs_scaled),test_imgs_scaled.shape[1]*test_imgs_scaled.shape[2])

._.DS_Store
565.jpg
997.jpg
766.jpg
982.jpg
170.jpg
366.jpg
.DS_Store
832.jpg
628.jpg
600.jpg
276.jpg
458.jpg
857.jpg
658.jpg
16.jpg
926.jpg
500.jpg
139.jpg
531.jpg
34.jpg
902.jpg
._222.jpg
._1.jpg
709.jpg
395.jpg
222.jpg
1.jpg
84.jpg
._.DS_Store
89.jpg
216.jpg
404.jpg
148.jpg
979.jpg
.DS_Store
588.jpg
._662.jpg
896.jpg
116.jpg
842.jpg
259.jpg
11.jpg
662.jpg
282.jpg
532.jpg
331.jpg
457.jpg
802.jpg
751.jpg
368.jpg
632.jpg
959.jpg
1.jpg
999.jpg
46.jpg
184.jpg
450
450 450
(400, 150, 150)


/Users/eslemus/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


In [12]:
from sklearn.metrics import confusion_matrix
y_pred = logisticRegr.predict(x_test_new)
cm = confusion_matrix(test_y, y_pred)

print("Accuracy Score -> ",accuracy_score(test_y, y_pred)*100)
print(classification_report(test_y,y_pred))
print(cm)

#print(y_pred)


Accuracy Score ->  88.44444444444444
              precision    recall  f1-score   support

           0       0.85      0.94      0.89       225
           1       0.93      0.83      0.88       225

    accuracy                           0.88       450
   macro avg       0.89      0.88      0.88       450
weighted avg       0.89      0.88      0.88       450

[[211  14]
 [ 38 187]]


Conclusion: When we spot check Logistic Regression optimization, there is an accuracy score of 88.44%